### PR TITLE
Residual Smoothing for IDR(s) solver.

### PR DESCRIPTION
**Pull request by RomainBrault**
Hello,

I've added an optional residual smoothing for the Induced Dimension Reduction method. It just follows the algorithm already implemented in python: https://github.com/astudillor/idrs_solver

Br,

Romain.

### CODE REVIEW

File: src/idrs.jl [hunk 1]
```diff
@@ -80,16 +81,23 @@ idrs(A, b; s = 8, tol=sqrt(eps(typeof(real(b[1])))), maxiter = length(b)^2) =
 idrs!(x, A, b; s = 8, tol=sqrt(eps(typeof(real(b[1])))), maxiter=length(x)^2) =
     idrs_core!(x, linsys_op, (A,), b, s, tol, maxiter)
 
-function idrs_core!{T}(X, op, args, C::T, s, tol, maxiter)
+function idrs_core!{T}(X::T, op, args, C::T, s::Int64, tol::Float64, maxiter::Int64; smoothing::Bool=false)
```
Comment (mschauer) [inline src/idrs.jl]: ``X::T`` is overspecialization. ``C`` is of type ``op(X, args...)``, so X and C not necessarily of the same type.

File: src/idrs.jl [hunk 2]
```diff
@@ -80,16 +81,23 @@ idrs(A, b; s = 8, tol=sqrt(eps(typeof(real(b[1])))), maxiter = length(b)^2) =
...
     normR = vecnorm(R)
     res = typeof(tol)[normR]
-	iter = 0                 
+	iter = 0
+
+    if smoothing
+        X_s = copy(X)
+        R_s = copy(R)
+        T_s = zeros(R)
+        # gamma = zeros(size(R_s, 1), size(T_s, 1))
```
Comment (mschauer) [inline src/idrs.jl]: delete this outcommented line?

File: src/idrs.jl [hunk 3]
```diff
@@ -144,41 +152,56 @@ function idrs_core!{T}(X, op, args, C::T, s, tol, maxiter)
...
             beta = f[k]/M[k,k]
             axpy!(-beta, G[k], R)
             axpy!(beta, U[k], X)
-        
+
             normR = vecnorm(R)
+            if smoothing
+                copy!(T_s, R_s); axpy!(-1., R, T_s) # T_s = R_s - R
+                gamma = vecdot(R_s, T_s)/vecnorm(T_s)^2
+                axpy!(-gamma, T_s, R_s) # R_s = R_s - gamma*T_s
```
Comment (mschauer) [inline src/idrs.jl]: This is very similar to the omega function, maybe a better abstraction? @astudillor

**Comment by mschauer** [comment]
Hi, thank you! I made some inline remarks. There is a testing error, can you look into this?
```
ERROR: LoadError: MethodError: `idrs_core!` has no method matching idrs_core!(::Array{Float32,1}, ::IterativeSolvers.#linsys_op, ::Tuple{Array{Float32,2}}, ::Array{Float32,1}, ::Int64, ::Float32, ::Int64)
```

**Comment by RomainBrault** [comment]
Hello,

I've already changed the code so that the test should pass (it was typing error). I'll push this version soon and then look at code factorisation!

**Comment by mschauer** [comment]
Almost, it really should be ``function idrs_core!{T}(X, op, args, C::T,...)`` so lines as ``P = T[rand!(copy(C)) for k in 1:s]´´ work correctly (note for myself: add a test to make this clear). Can you add a test to https://github.com/JuliaLang/IterativeSolvers.jl/blob/master/test/runtests.jl#L123 which exercises the new functionality?

**Comment by RomainBrault** [comment]
Ah sorry, I'll work on it tomorrow on the tests and and correct the types!

**Comment by mschauer** [comment]
Again, thank you for this PR.  Good to go as far as I'm concerned.

**Comment by mschauer** [comment]
``@collaborators``: Please comment or merge

**Comment by jiahao** [comment]
Squashed, cleaned and merged in #89. Thanks!

**Pull request closed without merging**